crates/accelerate/src/sabre/route.rs

```diff
@@ -0,0 +1,584 @@
+        self.front_layer.apply_swap(swap);
+        self.extended_set.apply_swap(swap);
+        self.layout.swap_physical(swap[0], swap[1]);
+    }
+
+    /// Return the node, if any, that is on this qubit and is routable with the current layout.
+    #[inline]
+    fn routable_node_on_qubit(&self, qubit: PhysicalQubit) -> Option<NodeIndex> {
+        self.front_layer.qubits()[qubit.index()].and_then(|(node, other)| {
+            self.target
+                .coupling
+                .contains_edge(NodeIndex::new(qubit.index()), NodeIndex::new(other.index()))
+                .then_some(node)
+        })
+    }
+
+    /// Update the system state as the given `nodes` are added to the routing order, preceded by
+    /// the given `swaps`.  This involves updating the output values `gate_order` and `out_map`,
+    /// but also the tracking objects `front_layer`, `extended_set` and `required_predecessors` by
+    /// removing the routed nodes and adding any now-reachable ones.
+    fn update_route(&mut self, nodes: &[NodeIndex], swaps: Vec<[PhysicalQubit; 2]>) {
+        // First node gets the swaps attached.  We don't add to the `gate_order` here because
+        // `route_reachable_nodes` is responsible for that part.
+        self.out_map
+            .insert(self.dag.dag[nodes[0]].py_node_id, swaps);
+        for node in nodes {
+            self.front_layer.remove(node);
+        }
+        self.route_reachable_nodes(nodes);
+        // Ideally we'd know how to mutate the extended set directly, but since its limited size
+        // ties its construction strongly to the iteration order through the front layer, it's not
+        // easy to do better than just emptying it and rebuilding.
+        self.extended_set.clear();
+        self.populate_extended_set();
+    }
+
+    /// Search forwards in the DAG from all the nodes in `nodes`, adding them to the `gate_order`
+    /// or the current `front_layer` as appropriate, and continue inspecting gates until there is
+    /// nothing further with no required predecessors.
+    ///
+    /// The nodes in `nodes` should all already have no further required predecessors.
+    fn route_reachable_nodes(&mut self, nodes: &[NodeIndex]) {
+        let mut to_visit = nodes.to_vec();
+        let mut i = 0;
+        let dag = &self.dag;
+        // Iterate through `to_visit`, except we often push new nodes onto the end of it.
+        while i < to_visit.len() {
+            let node_id = to_visit[i];
+            let node = &dag.dag[node_id];
+            i += 1;
+
+            // If the node is a directive that means it can be placed anywhere.
+            if !node.directive {
+                if let Some(blocks) = dag.node_blocks.get(&node.py_node_id) {
+                    let block_results = blocks
+                        .iter()
+                        .map(|block| self.route_control_flow_block(block))
+                        .collect::<Vec<_>>();
```

Comment:
Previously, the logic of `route_control_flow_block` was inlined here inside a `for` loop, with only the "swap epilogue" part encapsulated into a separate call, so it had ended up really nested, and it felt like the function calls didn't represent atomic operations.